Keep VRAM usage and faster slicing consistent in attention.py

mh-dm opened the pull request:
EDIT:
Refactor attention.CrossAttention to remove 4 copies of einsum_op_compvis code.
    
Remove zeros tensor that's useless when we have enough VRAM.
For cuda keep VRAM usage and faster slicing consistent.

mh-dm commented:
FYI @Any-Winter-4079 since you edited this recently.

i3oc9i commented:
I have tested thei PR on MacOs

512*512. -s50  /  **OK** but there is not improvment in speed
896*576 -s50 / **OK** but there is not improvment in speed
1024*1024 -s50 / **FAIL** with error Error: product of dimension sizes > 2**31'

Any-Winter-4079 commented:
Yeah, this is going to be tricky :)

For ongoing development of the solution, I would seek a thumbs up from 8GB (@Vargol), 16GB (@netsvetaev, @krummrey), 32GB (@0t0m0, @jroxendal) and 64 (@Any-Winter-4079) or 128GB (@i3oc9i) M1 machines before merging, as they seem to report various results. It should maintain the same speed for 512x512 and not crash on larger images.

I do agree there is probably a unified way to do this, though. I'm trying to get Textual Inversion working on M1, but I'll try to take a look. Thanks!

Vargol commented:
Going to be a while before I can get around to testing, I notice a lot of the VRAM reduction fixes have been remove which concerns me, but I guess the move to functions means they should fall out of scope and get collected so it might not be a disaster.

0t0m0 commented:
@Any-Winter-4079 
Thumbs down from me :(
The 512x512 image was created without problems in the normal time.
However, the 1024x1024 image took half the expected time and was colorful noise afterwards.

```
dream> banana sushi -W512 -H512 -Ak_lms -S50
100%|███████████████████████████████████████████| 50/50 [00:27<00:00,  1.83it/s]
Generating: 100%|█████████████████████████████████| 1/1 [00:31<00:00, 31.24s/it]
>> Usage stats:
>>   1 image(s) generated in 31.37s
Outputs:
[2] outputs/img-samples/000115.50.png: "banana sushi" -s50 -W512 -H512 -C7.5 -Ak_lms -S50

dream> banana sushi -W1024 -H1024 -Ak_lms -S50
>> This input is larger than your defaults. If you run out of memory, please use a smaller image.
100%|███████████████████████████████████████████| 50/50 [07:34<00:00,  9.09s/it]
Generating: 100%|████████████████████████████████| 1/1 [07:52<00:00, 472.54s/it]
>> Usage stats:
>>   1 image(s) generated in 472.67s
Outputs:
[3] outputs/img-samples/000116.50.png: "banana sushi" -s50 -W1024 -H1024 -C7.5 -Ak_lms -S50
```

jroxendal commented:
Non-scientific findings (i.e I shut down some of the worst memory hogs but didn't do a full restart to clear up memory)
32GB m1 max:
512x512 same as development, so about 35s.
896x576 hovering around 20s/it.
1024x1024 suspiciously fast at around 5s/it and generated colourful noise.

Vargol commented:
Sorry, can't test, seem to be in dependancy hell, I think something has been upgraded to sue a newer version of protobuff
that incompatible with a bunch of other dependancies

netsvetaev commented:
> 
> > For ongoing development of the solution, I would seek a thumbs up from 8GB (@Vargol), 16GB (@netsvetaev, @krummrey), 32GB (@0t0m0, @jroxendal) and 64 (@Any-Winter-4079) or 128GB (@i3oc9i) M1 machines before merging, as they seem to report various results. It should maintain the same speed for 512x512 and larger images (e.g. 1024x1024) and not crash on the latter.
> 
> My problems with ram may be related to non-arm brew and installation process differences. So I need to reinstall it. Right now fresh install do cause some errors though.

After a fresh install I've got 1.38s/it 512, 6.20s/it 768 and 20.5s/it 1024 (main branch). So it were mine problems with something non-arm.

netsvetaev commented:
> It should maintain the same speed for 512x512 and larger images (e.g. 1024x1024) and not crash on the latter.

Development:
512 at 1.04it/s (less than a second per it). 50s total.
768 at 6.20. 1024 at 22-24s/it. 

**This attention.py version:**
512 at 1.04s/it (slower)
768 at 10s/it
1024 had crashed dream.py with this:
```
/AppleInternal/Library/BuildRoots/5381bdfb-27e8-11ed-bdc1-96898e02b808/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShaders/MPSCore/Types/MPSNDArray.mm:705: failed assertion `[MPSTemporaryNDArray initWithDevice:descriptor:] Error: product of dimension sizes > 2**31'
Abort trap: 6
/Users/artur/miniconda3/envs/ldm10/lib/python3.10/multiprocessing/resource_tracker.py:224: UserWarning: resource_tracker: There appear to be 1 leaked semaphore objects to clean up at shutdown
  warnings.warn('resource_tracker: There appear to be %d '
```

lstein commented:
Tag me when this is ready for review. You might want to convert this PR into a draft if you're still actively working on it.

Vargol commented:
Sorted my dependency hell.

512x512, around normal speed
1024x1024 was running little slow (67s/it cf 55s/it) then crashed in 35th sample
	Insufficient Memory (00000008:kIOGPUCommandBufferCallbackErrorOutOfMemory)

mh-dm commented:
Thank you for testing this!
I see there's an issue with 1024x1024 / large sizes product of dimension sizes > 2**31 for mps (note 1024x1024 runs fine on just cpu) and I'll address that (looks like it was handled through `slice_size = math.floor(2**30 / (q.shape[0] * q.shape[1])` before my change).
@netsvetaev Surprised about any slowdown. This change is mostly a refactoring, especially for 512x512 where it should trigger the same algorithm with no slicing. I read you have 16GB, so for 768x768 size it should use larger slices. I'm guessing larger slices are slower, which is okay as I have to limit them to fix the above anyway.

krummrey commented:
@netsvetaev I'm in the office right now, will try it later this evening.

Vargol commented:
Darn, just realised that means I'm getting those memory issues from main, wasn't getting any from the 1.14.3 test repo.

i3oc9i commented:
@mh-dm 
after [force-pushed](https://github.com/lstein/stable-diffusion/compare/297e70d103e75899be5079f3b339c9620077b77c..b93338a0bf971ea29d5ab3fca186365701c150d2) the attention branch from [297e70d ](https://github.com/lstein/stable-diffusion/commit/297e70d103e75899be5079f3b339c9620077b77c)to [b93338a ](https://github.com/lstein/stable-diffusion/commit/b93338a0bf971ea29d5ab3fca186365701c150d2)

**banana split -s50 -W1024 -H1024**  works but it is almost two times slower than 1.14.1 release (2m55s)
```
50/50 [05:43<00:00,  6.87s/it]
1 image(s) generated in 345.72s
```
**banana split -s50 -W576 -H896** works and it s just a bit slower (50s in 1.14.1 relese)
```
50/50 [00:54<00:00,  1.09s/it]
1 image(s) generated in 55.06s
```
**banana split -s50 -W512 -H512** works at the same speed of 1.14.1 relese
```
50/50 [00:22<00:00,  2.25it/s]
1 image(s) generated in 22.83s
```

Vargol commented:
On, I can confirm in now for me 512x512 is **400% slower**

i3oc9i commented:
There is not lscpu in MacOs M1 sorry

krummrey commented:
Can someone give me a hand, how do I check out a pull request?
`git pull origin pull/582/head:development` didn't work.

mh-dm commented:
Maybe just `git pull origin pull/582/head`?

krummrey commented:
I ran it 3 time each (-n3):
512x512  -  `1:02`, `1:04`, `1:04`
640x640  -  `3:33`, `3:39`, `3:27`
No improvement here, even a slight increase in render times.

mh-dm commented:
Thank you for the test. Were you on b93338a or ac32c03 ?

mh-dm commented:
One small off by one error fixed and using smaller max tensor sizes by default. I think smaller sizes help fit some operations in L2/L3 caches which is why there was a perf difference before, or at least that's my best theory from the numbers you've kindly provided. If I limit to 32MiB then I get an almost 20% speed increase on intel ~10s to ~8s/it.
Can anyone on M1 give it a try?

Current max tensor size limit logic for mps, could be easily tuned up or down based on your testing:
```
        # mps have larger caches. But don't go too high to avoid:
        # 'large sizes product of dimension sizes > 2**31' error on mps
        if self.mem_total_gb <= 8:
            return self.einsum_op_tensor_mem(q, k, v, 64)
        elif self.mem_total_gb <= 16:
            return self.einsum_op_tensor_mem(q, k, v, 128)
        else:
            return self.einsum_op_tensor_mem(q, k, v, 256)
```

Any-Winter-4079 commented:
@mh-dm `sysctl -a` displays a lot of info on Mac.
For M1 Max 64GB
`sysctl -a | grep hw.l1`
```
hw.l1icachesize: 131072
hw.l1dcachesize: 65536
```
Those are L1 instruction and data cache sizes
`sysctl -a | grep hw.l2`
```
hw.l2cachesize: 4194304
```
It doesn't appear to return any info on L3, so maybe there's not one (?).

Vargol commented:
> Can someone give me a hand, how do I check out a pull request?
> `git pull origin pull/582/head:development` didn't work.


Checkout a pull request
% git fetch origin pull/582/head:test_582
% git checkout test_582    

Thats what I've been using

krummrey commented:
Checked it out the Vargol way and ran it 3 time each (-n3):
512x512 - `1:03`, `1:04`, `1:03`
640x640 - `3:19`, `3:47`, `3:23`

jroxendal commented:
Ok, really promising results from latest commits: 
512x512 same as usual.
1024x384 1m14s, about 15seconds faster than lstein:development
1024x1024 took about 9 minutes, but this is the first time I'm completed generation at that size! no visible artefacts. memory pressure didn't even hit the ceiling, I think I could go every higher. Very promising indeed.

Vargol commented:
@jroxendal

Whats your set up, been doing 1024x1024 on a 8GB shared memory Apple M1 for a while, although the release version of 1.14 seems to have issues the 1.14.3 performance test version didn't in that regard

Vargol commented:
Still 400% slower here at 512x512

mh-dm commented:
@jroxendal Thank you. First speedup confirmation 🎉, on 32GB M1 max. From code that should be limiting tensor size to 256MiB
```python
return self.einsum_op_tensor_mem(q, k, v, 256)
```
@krummrey 16GB M1 basic right? I don't have baseline (`git checkout origin development`) numbers from you, so I can't tell whether slower or faster. But the high variability for 640x640 tells me to go down from 128 to 64MiB max tensor size.

> Still 400% slower here at 512x512 and use 2 -3 more GB than the release version.
@Vargol This is surprising given the code switch to 64MiB max tensor size for <=8GiB. The higher ram usage is indicative of using an older version (the one with 1024), maybe a quick refetch & checkout and make sure you're on commit 2df118d? You could also try replacing the whole einsum_op_cpu like
```python
def einsum_op_cpu(self, q, k, v):
  return self.einsum_op_tensor_mem(q, k, v, 64)
```
And try different values like 64, 32, 24, 16.

mh-dm commented:
@Any-Winter-4079 Thank you for that info. sysctl -a doesn't show L* cache sizes on linux unfortunately so we're still missing a common way of getting the values. Could you give the latest code, 2df118d, in this PR a try on 512x512?

Has anyone tried `getconf -a | grep CACHE_SIZE`?

jroxendal commented:
@mh-dm I'm afraid I have to retract my speedup report from before: 
development branch 1024x1024 is at ~8s/it and this branch at 12s/it, so slower. Pardon the confusion.

mh-dm commented:
@jroxendal Thanks. How's 512x512 and some size in between like 640x640?

Any-Winter-4079 commented:
> @Any-Winter-4079 Thank you for that info. `sysctl -a` doesn't show L* cache sizes on linux unfortunately so we're still missing a common way of getting the values. Could you give the latest code, [2df118d](https://github.com/lstein/stable-diffusion/commit/2df118d4641960296e112247e3f99d9052bd99cc), in this PR a try on 512x512 and maybe 640/768?
> 
> Has anyone tried `getconf -a | grep CACHE_SIZE`?

`getconf -a | grep CACHE_SIZE` doesn't work on M1, but you can always get the data in different ways based on the device type, e.g. `if torch.cuda.is_available(): do X else do Y`

jroxendal commented:
512x512 [00:33<00:00,  1.47it/s] dev
512x512  [00:35<00:00,  1.40it/s] this branch 
640x640 [01:01<00:00,  1.23s/it] dev
640x640 [01:17<00:00,  1.55s/it] this branch

mh-dm commented:
@Any-Winter-4079 Thank you. I know I can switch on device type, but I was looking to switch on more granular information. If L3_cache_size_mb <= 8: split at 32 else split at 64

Vargol commented:
You'll need to take into account that your CPU ops methods are currently dealing with M1 'APUs' and intel CPU with AMD graphics cards, they all fall under the 'mps' device type as its METAL based not M1 specific.

mh-dm commented:
@jroxendal Thank you.
Okay, time to pivot. I will undo and undo parts of this change until it will be a pure refactoring for mps.

mh-dm commented:
I've brought back `einsum_op_mps_v1` and `einsum_op_mps_v2` and the exact same logic to pick between the two. Also rebased on top development head. Could anyone test on M1* to confirm that speeds for say 512x512 and 640x640 are unchanged compared to development head?

Vargol commented:
@Any-Winter-4079  beat me to it, currently testing with the >= fixed to > than as without the fix 1024x1024 is  4 MINUTES per it

mh-dm commented:
@Any-Winter-4079 Not sure that typo matters in practice as RAM nowadays isn't exactly integer GiBs so it won't divide equally. Regardless the issue is in main and this PR no longer changes that.

Vargol commented:
> @Any-Winter-4079 Not sure that typo matters in practice as RAM nowadays isn't exactly integer GiBs so it won't divide equally. Regardless the issue is in development head and this PR no longer changes that.

Yes it matters ... Some one with a 8Gb M1 :-)
As your code has that line in too so if I was to test this PR as is ir would have a great big thumbs from me.

mh-dm commented:
> Yes it matters ... Some one with a 8Gb M1 :-) Your code has that line in too so if I was to test this PR 'as is' it would have a great big thumbs down from me.

8GB is different from 8GiB which is probably different from what psutil.virtual_memory().total will report if size isn't exact or some memory is reserved. For example, I have 32GB memory, but:
```python
>>> import psutil; print(psutil.virtual_memory().total / (1024**3))
31.121078491210938
```
Try it as an easy way to check if the `=` in `self.mem_total >= 8` makes a difference or not.

Vargol commented:
I don't need to try it

>= 8      512x512   @ 25 s/it
> 8        512x512   @   5 s/it

mh-dm commented:
Thanks for checking. Are the numbers for the latest version of this pull request?

Vargol commented:
As long as you haven't updated it since
https://github.com/lstein/stable-diffusion/pull/582#issuecomment-1249290586
then yes its with the latest

mh-dm commented:
Great, then including switch to self.mem_total > 8  in this PR.

Any-Winter-4079 commented:
> > Yes it matters ... Some one with a 8Gb M1 :-) Your code has that line in too so if I was to test this PR 'as is' it would have a great big thumbs down from me.
> 
> 8GB is different from 8GiB which is probably different from what psutil.virtual_memory().total will report if size isn't exact or some memory is reserved. For example, I have 32GB memory, but:
> 
> ```python
> >>> import psutil; print(psutil.virtual_memory().total / (1024**3))
> 31.121078491210938
> ```
> 
> Try it as an easy way to check if the `=` in `self.mem_total >= 8` makes a difference or not.

Interesting. That is probably a difference with Macs. Here it's exact values, so 64.0 means 64GB Mac and so on.
```
>>> psutil.virtual_memory().total
68719476736
>>> psutil.virtual_memory().total / (1024**3)
64.0
```

Vargol commented:
Yep, it's strange that someone decided to start applying fake disk storage units to memory as while, I'm sure it saves someone a penny a memory stick or something.

i3oc9i commented:
test on commit bcfa635aef35b55d902c43d6b8302db79f009513

**"banana split" -s50 -W1024 -H1024 -C7.5 -Ak_lms**  
```
[02:52<00:00, 172.18s/it]
1 image(s) generated in 172.37s
```
just 3 sec faster than last development branch 45af30f3a4c98b50c755717831c5fff75a3a8b43
**FAIL** because I get noise (did two runs to cofirm)

**"banana split" -s50  -W896 -H576 -C7.5 -Ak_lms**  
```
50/50 [00:53<00:00,  1.07s/it] 
1 image(s) generated in 54.10s
```
**OK** same execution speed

**"banana split" -s50  -W512 -H512 -C7.5 -Ak_lms**  
```
50/50 [00:21<00:00,  2.31it/s]
1 image(s) generated in 22.19s
```
**OK** same execution speed

Vargol commented:
512x512 ... 👍 

100%|████████████████| 50/50 [03:51<00:00,  4.63s/it]
100%|████████████████| 50/50 [04:00<00:00,  4.80s/it]
100%|████████████████| 50/50 [04:02<00:00,  4.85s/it]
Generating: 100%|████████████████████████████████████| 3/3 [12:28<00:00, 249.49s/it]
>> Usage stats:
>>   3 image(s) generated in 748.69s


1024x1024 looking good speed wise (despite it throwing out an error everyone again, nothing to do with this change though)
but it'll be 2.5 hour before I can confirm cos thats how slow it is for me :-)

mh-dm commented:
@i3oc9i Thank you. Hmm.. I'm assuming you don't get noise at dev head for 1024. I did another self review and checked for any logic `einsum_op_mps_v1` in `einsum_op_mps_v2` that might result in any difference. One I see is `s = einsum('b i d, b j d -> b i j', q, k) * self.scale` vs `s = einsum('b i d, b j d -> b i j', q, k); s *= self.scale`. That would be weird if it were the cause, but that gave me an idea on how to get a clear speedup: moving `* self.scale` to before `einsum()`, specifically to `k` that's most often both smaller than `q` and always much smaller than `s`.

Would appreciate further testing.

Vargol commented:
that actually might be it, I can't test it of course as I never hit the CompVis op, however it looks similar to this commit that was done to back out issues when training

https://github.com/lstein/stable-diffusion/commit/bf1beaa607984e34f3bfab8b264444318b6f47b0

In that case a bunch of `+=` thad to be reverted back to addition.

i3oc9i commented:
@mh-dm 
test on commit 1114da7b072e236c6f84deb00923485699e9fe0b

"banana split" -s50 -W512 -H512 -C7.5 -Ak_lms. (OK same speed)
"banana split" -s50 -W896 -H576 -C7.5 -Ak_lms (OK a bit faster)
"banana split" -s50 -W1024 -H1024 -C7.5 -Ak_lms (OK 2:55 😀 You got it 👏)

mh-dm commented:
@i3oc9i Awesome 🎉!

I think I can call it ready for review :)

Vargol commented:
```
    def einsum_op_mps_v2(self, q, k, v):
        if self.mem_total > 8 and q.shape[1] <= 4096:
            return self.einsum_op_compvis(q, k, v)
        else:
            return self.einsum_op_slice_0(q, k, v, 1)  <-- ****I always go here****

```

mh-dm commented:
Yeah and `einsum_op_slice_0` calls `einsum_op_compvis` for each slice.

netsvetaev commented:
Guys, I'm on 16gb m1. 
Right now on main I have 1.3/512 at minimum, but 5.8-6.2s/768, 19s/1024 (~15m total); 
on dev 1.04it/512 (+30% faster), but 6.7s/768 (slower), 26s/1024 (slower, 21m total).

P.S. «q» doesn't work on dev.

Vargol commented:
> Yeah and `einsum_op_slice_0` calls `einsum_op_compvis` for each slice.


Then I apologise,I must have been  remembering the code wrong

netsvetaev commented:
> @netsvetaev Could you please compare this PR to development? Not sure what «q» is?

Sorry, I thought this is the development. How I can compare then?
«q» — command for quit dream.py.

mh-dm commented:
```sh
git fetch origin pull/582/head:test_582
git checkout test_582
```
Now you're on this PR and you can test.
```sh
git checkout HEAD^
```
Now you're on development (since this PR is only one commit) at whatever this PR was last synced to.

If you just want latest development:
```
git checkout origin/development
```

netsvetaev commented:
> @netsvetaev Could you please compare this PR to development? 

1.05it/512
6.23s/768 (jumps from 5.7 to 6.7)
23.5s/1024

Any-Winter-4079 commented:
`"banana sushi" -s50 -C7.5 -n3 -W512 -H512`
3 image(s) generated in 81.13s (86s previously) 

`"banana sushi" -s50 -C7.5 -n3 -W1024 -H1024`
3 image(s) generated in 716.92s (782s previously)

Big thumbs up from me @mh-dm . Having debug prints and other stuff running, and even then, best times.

Any-Winter-4079 commented:
Do we have the thumbs up from everyone from 8 to 64/128GB?

64 and 128GB is a 👍

netsvetaev commented:
> Do we have the thumbs up from everyone from 8 to 64/128GB?
> 
> 64 and 128GB is a 👍

I see some slowdown when there is not enough memory. Maybe it’s a random case. 512 seems faster a bit, 768 more or less the same, 1024 somewhat sower. It would be great if there was a way to deal with it.

krummrey commented:
I ran it 3 time each (-n3):
512x512 - 1:14, 1:06, 1:12
640x640 - 3:03, 2:45, 2:39

512x512 went right into swap which it didn't do before, I guess thats where the increased rendering time comes from. (when I tested v1.14 I had some renders under a minute. so 10-15% faster back then)
640x640 is faster now.

Vargol commented:
I ran a proper 1024x1024 overnight and the results where in the 60 - 66 s/it region, which is in the right area and I get very inconstant values at that size anyway, probably use to OS things kicking in and stealing bits of memory.

512x512 is as fast as usual

-W704 -H512 was slow 1709 vs 1299 for my best run, I need to test against development to see what happen, but thats not going to happen for a few hours.

Any-Winter-4079 commented:
`"banana sushi" -s50 -C7.5 -n3 -W704 -H512` is faster for me (64GB)
3 image(s) generated in 124.35s (133s before)
Just to report at some RAM sizes, the current code is faster at that size too.

mh-dm commented:
It's going to be really hard for me to deliver mps specific optimizations, as somewhat experienced in this thread :). Best I can do is look for architecture independent optimizations (like moving `* self.scale` earlier on a smaller tensor) and maintain existing architecture specific optimizations while refactoring, like how I've brought back `einsum_op_mps_v1` and `einsum_op_mps_v2` and the logic to switch between them.

Extra difficulty comes from variability while timing/profiling and external factors. Just an example, I've found out that which power supply I use significantly affects my results on cpu - 12s/it to 10s/it (timed both on development). That's on top of known other factors like: amount free RAM, any other apps running, whether plugged in or not and performance profiles.

So, I've updated the code to more cleanly separate the architectures, specifically to separate paths for `mps` and `cpu` (where I could now test and apply a smaller&faster max tensor size). I've also rebased to top of development head. With everything refactored it should be easier for someone on mps to go in and try out future optimizations. Maybe it works to switch to `self.mem_total > 16` in `einsum_op_mps_v2` or maybe to just call `einsum_op_tensor_mem(self, q, k, v, some_ideal_limit_mb)` or maybe something else.

With all the testing (thank you!) and many good numbers I'm hoping we can call it good enough soon. Also, would appreciate some fresh eyes on the code looking for any bugs left.

Any-Winter-4079 commented:
@mh-dm Let me say I appreciate your efforts. You are clearly more knowledgeable than I am, but still it seems this issue is a pain in the ass :)

The optimization works great for 64GB and 128GB: For me at 64GB, it's 10% less time al 1024x1024 and 6% less time on 512x512 so that's pretty significant for me! 👍  If 16-32GB report worse results, we can always let them use the current dev version (meaning, just route them to the existing function -which is very simple to do, because reported RAM on Mac is exact.
So here `self.einsum_op = self.einsum_op_mps_v1 if self.mem_total >= 32 else self.einsum_op_mps_v2` we can route them to the existing version if they prefer so. Anyway, thanks again!

i3oc9i commented:
on last commit 654ab4eb5d2ad7201d3ebf091810f51de459a97a

"banana split" -s50 -W512 -H512 -C7.5 -Ak_lms. (**OK** 0:21)
"banana split" -s50 -W896 -H576 -C7.5 -Ak_lms (**OK** 0:50 )
"banana split" -s50 -W1024 -H1024 -C7.5 -Ak_lms (**OK** 2:44)

mh-dm commented:
"we can always let them use the current dev version (meaning, just route them to the existing function -which is very simple to do, because reported RAM on Mac is exact)"
If you look at the code, that's basically what it's doing. @Vargol reported okay for 512x512 on 8GB and for that specifically I did the `>= 8` to `> 8` change requested. It's only @krummrey on 16GB that's reporting a slowdown. That slowdown I don't understand at all because all the splitting logic is the same. However, netsvetaev@ also on 16GB reports "512 seems faster a bit, 768 more or less the same, 1024 somewhat slower". I believe further optimizations/changes specific to 16GB should be in a future change worked on by someone with access to said machine.

Vargol commented:
I'm still around an hour  from confirming for definite but my slow down at 
-W704 -H512  seems to be inherent to the dev version, still need to check main

For my old fork, 100 odd commits behind main 
>> Usage stats:
>>   3 image(s) generated in 1244.17s

For the development branch, checked out and pulled to make sure its up to date
>> Usage stats:
>>   3 image(s) generated in 1629.39

The development branch, replacing attention.py and model.py with copies from my fork...
Not finished yet but the first iter was 10.5s/it very similar to the dev branch, the
second is 50% though at also showing ~10.5s/it

mh-dm commented:
@Vargol You're getting a "400%" speed up for 512x512 from this PR, i.e. the change from `>= 8` to `> 8` :D, compared to development. Don't think there's more this PR can do to top that :).

mh-dm commented:
I re-opened review but I forgot to re-tag @lstein

Any-Winter-4079 commented:
If you compare 64GB vs. 128GB:

`"banana split" -s50 -W512 -H512 -C7.5 -Ak_lms`. (OK 0:26)
Comparison with 2x RAM: 24% slower
`"banana split" -s50 -W896 -H576 -C7.5 -Ak_lms` (OK 1:10)
Comparison with 2x RAM: 40% slower
`"banana split" -s50 -W1024 -H1024 -C7.5 -Ak_lms` (OK 3:44)
Comparison with 2x RAM: 37% slower

If you compare (best from) 16GB vs. 128GB:

`"banana split" -s50 -W512 -H512 -C7.5 -Ak_lms`. (OK 0:48)
Comparison with 8x RAM: 85% slower
`"banana split" -s50 -W896 -H576 -C7.5 -Ak_lms` (OK 5:12)
Comparison with 8x RAM: 624% slower
`"banana split" -s50 -W1024 -H1024 -C7.5 -Ak_lms` (OK 19:35)
Comparison with 8x RAM: 716% slower

Clearly, it's not just GPU cores that matter, but there's probably swap going on. So in the future, to optimise 16GB even further, one would have to probably look at that. But without a 16GB Mac like @mh-dm said, it's hard to optimise.

Vargol commented:
Yes, ignoring that the the dev branch was a 400% slow down from main due to that typo:-)

But the point of that post was to prove to myself, even if no one else, that the issues I'm seeing with '-W704 -H512 ' is nothing to do with this PR.

Any-Winter-4079 commented:
@mh-dm Yep, you were right that it is almost the exactly same code for 16GB.
This returns your memory, e.g. 16 for 16GB.

<img width="467" alt="Screenshot 2022-09-17 at 18 48 10" src="https://user-images.githubusercontent.com/50542132/190867654-f9a7e02f-9dd4-4076-971e-3c0119840354.png">

With 16GB, you'd go through `einsum_op_mps_v2`

<img width="337" alt="Screenshot 2022-09-17 at 18 48 46" src="https://user-images.githubusercontent.com/50542132/190867681-c6e57656-019b-4f25-ae5d-1fd44f137cdc.png">

For images 512x512, run `einsum_op_compvis`

<img width="359" alt="Screenshot 2022-09-17 at 18 57 54" src="https://user-images.githubusercontent.com/50542132/190868046-2c92c7d9-81bb-4a4e-85e4-662f14d18843.png">

which is

<img width="394" alt="Screenshot 2022-09-17 at 18 50 22" src="https://user-images.githubusercontent.com/50542132/190867745-36a9167e-0594-4a49-835d-fa36636cfdaa.png">

Else, use `einsum_op_slice_0` with `slice_size` 1

<img width="640" alt="Screenshot 2022-09-17 at 18 52 12" src="https://user-images.githubusercontent.com/50542132/190867828-f40a346f-47c1-4b70-9c53-30924812e904.png">

Compared to current dev branch code (with >=8 typo)

<img width="529" alt="Screenshot 2022-09-17 at 18 53 38" src="https://user-images.githubusercontent.com/50542132/190867888-0fd83aac-6d11-4f26-8306-f39544098fe6.png">

Maybe the available RAM at any point affects performance (e.g. one day may report faster results than a different day, because of other apps, etc. running).

If the different results from dev branch were from the same day, I'm not sure what it could be. Maybe calling the `einsum_op_compvis`  function with portions of the data, e.g.` r[i:end] = self.einsum_op_compvis(q[i:end], k[i:end], v[i:end])`  (which I don't know how it works -maybe the slice of data is copied using ranges?)

Other than that, I really don't see any difference.

mh-dm commented:
Thanks for looking. Can I get the github review approve so I can submit then?

Any-Winter-4079 reviewed (approved):
This commit gives 6-10% fasters results for 64-128GB (10% at larger sizes) M1 and fixes a typo for 8GB M1. Performance on 32GB M1 has not been tested (maybe I missed it in the Conversation), but will run the same code as 64-128GB, which is the behavior already present on the dev branch. For 16GB M1, there are mixed results (more testing would be needed).

Any-Winter-4079 commented:
> Thanks for looking. Can I get the github review approve so I can submit then?
> 
> Oh and to answer the question, slices like `q[i:end]` are not copies.

By the way, have you tested on CUDA?

mh-dm commented:
Yeah, tested ~6% faster on 8gb nvidia card for 512x512. I tested various sizes until 1024x1024 at which point it was ~10% faster. Intel v7 cpu got the highest improvement, I think ~25% overall for 512x512 down to 7-8s/it though.

Any-Winter-4079 commented:
@krummrey @netsvetaev 
By the way, because 16GB have reported mixed results, if anyone with 16GB RAM tests (same time of day, same apps open/closed, exiting dream> after every test, leaving a few minutes not to overheat -which is a real thing for me at 64GB impacting performance) and there seems to be a consistent slowdown, you can always do a Pull request to add your own code to run for 16GB Mac. For example here
<img width="413" alt="Screenshot 2022-09-17 at 19 25 49" src="https://user-images.githubusercontent.com/50542132/190869146-16bbed26-dac7-424e-95af-29a9a2e3fd5e.png">
Just need to add an if, e.g. `if self.mem_total_gb==16` and then just call the current code you are running (or an even more improved version). The current code (before this merge) is explained here. https://github.com/lstein/stable-diffusion/pull/582#issuecomment-1250106383 👍

mh-dm commented:
![image](https://user-images.githubusercontent.com/299015/190869656-5e295342-7bbe-4cf0-9482-809f240ddc3d.png)
One thing I'd try when experimenting with optimizations is changing `> 8` into `> 16` to always call `slice_0`.

Another thing to try would be:
```python
    def einsum_op_mps_v2(self, q, k, v):
        if self.mem_total_gb <= 8:
            return self.einsum_op_slice_0(q, k, v, 1)
        return self.einsum_op_tensor_mem(q, k, v, 64)
```
Where you try different values, `64`, `48`, `32`, `24`, `16`.

netsvetaev commented:
Seems it's got faster on a latest dev:
1.11it/s 512 (was 1.04 maximum), 45s total
5.25s/it 768 (was 6.20), 04:22s total
That's impressive!

For 1024 I have same numbers from 23s, increasing with every sample up to 25-26 (was 19s as best result earlier).

I will try to edit the code, thank you for the tips.

i3oc9i commented:
on last commit e040c35d34131c7ffcc55bd9c1ec8e21d5f48025

"banana split" -s50 -W512 -H512 -C7.5 -Ak_lms. (**OK** 0:20) same as previous report
"banana split" -s50 -W896 -H576 -C7.5 -Ak_lms (**OK** 0:50 )  same as previous report
"banana split" -s50 -W1024 -H1024 -C7.5 -Ak_lms (**OK** 2:44) same as previous report

Birch-san commented:
I've tried this out on my fork:  
https://github.com/Birch-san/stable-diffusion/compare/birch-mps-waifu...optimized-attention

using very recent PyTorch nightly: 1.13.0.dev20220917 on M1 Max (64GB).

### 1 sample:

**optimized:**  
11.4 secs

**non-optimized:**  
10.4 secs

### 3 samples:

**optimized:**  
37.8 secs

**non-optimized:**  
30.2 secs

### Conclusion

Optimized attention is still slower than "CompVis original code + some `del`s". for small batches and also large.

Vargol commented:
@Birch-san  Have you tried in with a release pytorch I see significant slowdown with the nightlies.

Birch-san commented:
using PyTorch stable 1.12.1

### 1 sample

**optimized**:  
10.1 secs

**non-optimized**:  
9.8 secs

### 3 samples

**optimized**:  
31.2 secs

**non-optimized**:  
28.7 secs

### Conclusion

Optimized attention is 3~9% slower than "CompVis original code + some `del`s".  
PyTorch stable is 5% faster than recent nightly.

Birch-san commented:
interesting, it seems I'm being sent down the M1 16–32GB code path. my `memory_available` was measured to be 31.89 GB. 😛

Birch-san commented:
using PyTorch stable 1.12.1

### 1 sample

**optimized**:  
9.8 secs

**non-optimized**:  
9.8 secs

### 3 samples

**optimized**:  
26.9 secs

**non-optimized**:  
28.7 secs

### Conclusion

If you go down the right code path (i.e. quitting apps to free up 32GB of memory), then optimized attention is identical speed for 1 sample, and 6.7% faster for a batch-of-3 samples.

Birch-san commented:
btw, I think I see a way to make inference faster.  
instead of allocating new tensors each time: we could retain and re-use tensors.  
it'd be a bit of a mess though. potentially increases the peak VRAM use (because you'd have a few large tensors that you'd never free).

I also wonder whether slapping `.contiguous()` on any of these tensors would make it cheaper to do arithmetic on them. due to improved locality.

mh-dm commented:
> btw, I think I see a way to make inference faster. instead of allocating new tensors each time: we could retain and re-use tensors. it'd be a bit of a mess though. potentially increases the peak VRAM use (because you'd have a few large tensors that you'd never free).

pytorch really doesn't like modifying/re-using tensors. I had to re-land #569 with a .clone() fix so that training still works and even a simple `*=  scalar` seemed to introduce issues for 1024x1024 on M1. Finally, there's not that much left to speed up unfortunately when the slowest operation/the bottleneck is `s = s.softmax(dim=-1)` for which there's no inplace version available in pytorch. If anyone reading has contributed to pytorch before and would like a nice rewarding challenge try adding support for `softmax_` or `softmax(inplace=True)`.

> I also wonder whether slapping `.contiguous()` on any of these tensors would make it cheaper to do arithmetic on them. due to improved locality.

Less chance of seeing significant improvements as you'd be introducing a copy. Plus I think reshape() already results in contiguous copy.

> interesting, it seems I'm being sent down the M1 16–32GB code path. my memory_available was measured to be 31.89 GB

You have the configuration to try out M1 16GB+ specific optimizations that Any-Winter-4079 and I have suggested, just a few messages up.

Any-Winter-4079 commented:
> interesting, it seems I'm being sent down the M1 16–32GB code path. my `memory_available` was measured to be 31.89 GB. 😛

> If you go down the right code path (i.e. quitting apps to free up 32GB of memory), then optimized attention is identical speed for 1 sample, and 6.7% faster for a batch-of-3 samples.

@Birch-san This new commit doesn't use mem_available on M1 though. It uses mem_total. You shouldn't need to quit apps nor get 31.89 of available memory (b/c it's mem_total that's measured).
